### DISCUSSION TITLE
Which backend profile for doing token exchange from a PKCE frontend issued token?

### DISCUSSION THREAD

**Discussion by danielloader** [zitadel/zitadel#9078] [Q&A] [answered]
I read the warning in the documentation:

> IMPORTANT
> 
> ZITADEL allows any application to use Token Exchange, however we strongly recommend to only configure confidential clients (using either client credentials or JWT assertion) with the Token Exchange grant type. This is because there is some trust placed in the application when it comes to defining scope and that it obtained tokens in a legitimate way. For example, if the app possesses a token of an admin user with impersonation permissions it can obtain tokens for any other user in your instance. It is your responsibility to make sure the application can be trusted with this kind of powers. If you configure a public client with the Token Exchange grant, you risk a leaked token can be used by an attacker who knows the client ID of a granted public client.

Which makes sense, I don't want my frontend to do token exchange, I want my frontend to go via a PKCE flow, store the opaque token in the session in the browser, and then offer an opaque token to my backends via HTTP API calls.

That said, I have integrations with third party software the backends need to call that require JWTs (apache superset for one) so I need to exchange these opaque tokens for JWTs somewhere - currently implementing a very simple golang service to do so that on hitting this proxy, it'll take the opaque token and pass along a JWT instead to services that can't use the opaque token. So far so good conceptually.

That said I can't work out how I'd configure a backend to do this: 

- If I pick the API Application template in a project and then the Basic auth method, it does give me the client id and secret I could use for a token exchange but the grant is not configurable to enable token exchange on this application.
- I can pick the Web Application template and code authentication method, and add the grant - but is that correct for a headless API service? What am I doing wrong if I do this? What's the risks of abusing this method since I don't need to generate sessions or tokens via this application, merely exchange valid ones from an existing application?

If I use a client credentials grant API application instead of the code web application, when I go to use the exchange endpoint (which works if I use a web app configured with token exchange in the grants) I get a somewhat not useful error message:

```json
{
  "error": "invalid_client",
  "error_description": "no active client not found"
}
```

**Comment by muhlemmer** [comment, marked as answer]
Hi @danielloader. 

> If I pick the API Application template in a project and then the Basic auth method, it does give me the client id and secret I could use for a token exchange but the grant is not configurable to enable token exchange on this application.

Yes, that is an older limitation in our datamodels where we can't store OAuth2/OIDC specific configuration on an API. Therefore we can't set grant types. The "type" of application only determines the configurable options. Your use case might be described as "backend for front-end" and thus a web-app. 

I recommend creating a web-app and enabling the grant type of token exchange.

We will take your feedback into account in improving our product in this sense. Thanks.

**Reply by danielloader** [reply]
That'll do, was more concerned I was just doing this all wrong and wanted some validation on the issue at hand.

Will experiment and go from there.
